File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/forbid-blank-line-in-leaf-function.ts
#!/usr/bin/env bun
import type { FunctionLike, Node, SourceFile } from "typescript@6.0.3";
import {
	createSourceFile,
	forEachChild,
	isBlock,
	isFunctionDeclaration,
	isIdentifier,
	isMethodDeclaration,
	SyntaxKind,
} from "typescript@6.0.3";
import type {
	Finding,
	HarnessCheckRule,
	HarnessManifest,
	RuleContext,
} from "../harness-check-rule";

/**
 * Helper to check if a function has nested functions.
 */
export const forbidBlankLineInLeafFunctionRule = (
	ctx: RuleContext,
): HarnessCheckRule => ({
	applies(_manifest: HarnessManifest): boolean {
		return true;
	},

	validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
		return ctx.stackSources(manifest, "typescript").flatMap((file) => {
			const text = ctx.read(file);
			if (!text) {
				return [];
			}

			let sourceFile: SourceFile;
			try {
				sourceFile = createSourceFile(
					file,
					text,
					SyntaxKind.LatestVersion,
					true,
				);
			} catch {
				return [
					{
						severity: ctx.severityOf(manifest, "forbidBlankLineInLeafFunction"),
						category: "forbidBlankLineInLeafFunction",
						message: `failed to parse TypeScript: ${file}`,
					},
				];
			}

			const findings: Finding[] = [];

			const extractBlankLineFindings = (
				funcNode: FunctionLike,
				body: FunctionLike["body"],
			): readonly Finding[] => {
				const blankLineFindings: Finding[] = [];
				if (!isBlock(body) || body.statements.length === 0) {
					return blankLineFindings;
				}

				const statements = body.statements;

				const funcName =
					(isFunctionDeclaration(funcNode) && funcNode.name?.text) ||
					(isMethodDeclaration(funcNode) &&
						funcNode.name &&
						isIdentifier(funcNode.name) &&
						funcNode.name.text) ||
					"<anonymous>";

				const checkTrivia = (triviaStart: number, triviaEnd: number): void => {
					const triviaStartLine =
						sourceFile.getLineAndCharacterOfPosition(triviaStart).line;
					blankLineFindings.push(
						...text
							.slice(triviaStart, triviaEnd)
							.split(/\r?\n/)
							.map((line, i) => ({ line, index: i }))
							.filter(({ line }) => line.trim() === "")
							.map(({ index }) => ({
								severity: ctx.severityOf(
									manifest,
									"forbidBlankLineInLeafFunction",
								),
								category: "forbidBlankLineInLeafFunction",
								message: `${file}:${triviaStartLine + index + 1}: leaf function \`${funcName}\` contains a blank line; remove or extract the section`,
							})),
					);
				};

				if (statements.length > 0) {
					checkTrivia(
						body.getStart(sourceFile, true),
						statements[0].getFullStart(),
					);
				}
				statements.slice(0, -1).forEach((stmt, i) => {
					checkTrivia(stmt.getEnd(), statements[i + 1].getFullStart());
				});
				if (statements.length > 0) {
					checkTrivia(
						statements[statements.length - 1].getEnd(),
						body.getEnd(),
					);
				}

				return blankLineFindings;
			};

			const visit = (node: Node): void => {
				switch (node.kind) {
					case SyntaxKind.FunctionDeclaration:
					case SyntaxKind.MethodDeclaration:
					case SyntaxKind.FunctionExpression:
					case SyntaxKind.ArrowFunction:
					case SyntaxKind.Constructor: {
						const funcLike = node as FunctionLike;
						if (funcLike.body && !hasNestedFunctions(funcLike)) {
							findings.push(
								...extractBlankLineFindings(funcLike, funcLike.body),
							);
						}
						break;
					}
				}
				forEachChild(node, visit);
			};

			visit(sourceFile);
			return findings;
		});
	},
});

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/forbid_blank_line_in_leaf_function.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Forbid blank line in leaf function rule."""

from pathlib import Path

import libcst as cst

from harness_check_rule import Finding, HarnessCheckRule

from ._utils import has_nested_function, parse_python, relative, severity_for, stack_sources


class ForbidBlankLineInLeafFunctionRule(HarnessCheckRule):
    """Validate forbidBlankLineInLeafFunction check."""

    category = "forbidBlankLineInLeafFunction"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        enabled = section.get("enabled", True)
        return enabled is not False

    def validate(self, root: Path, manifest: dict) -> list[Finding]:
        """Validate forbidBlankLineInLeafFunction check."""
        severity = severity_for(manifest, self.category)
        sources = stack_sources(root, manifest, self.category)
        class _BlankLineFinder(cst.CSTVisitor):
            def __init__(self, rel_path: str) -> None:
                super().__init__()
                self.findings: list[Finding] = []
                self.rel_path = rel_path
            def visit_FunctionDef(self, node: cst.FunctionDef) -> bool:
                if has_nested_function(node):
                    return False
                if not isinstance(node.body, cst.IndentedBlock):
                    return False
                func_name = node.name.value
                for stmt in node.body.body:
                    if isinstance(stmt, cst.EmptyLine) and stmt.comment is None:
                        pos = self.get_metadata(cst.metadata.PositionProvider, stmt)
                        self.findings.append(Finding(
                            severity,
                            self.category,
                            f"{self.rel_path}:{pos.start.line}: leaf function `{func_name}` contains a blank line; remove or extract the section",
                        ))
                return True
        result = []
        for path in sources:
            tree, error = parse_python(path)
            if error is not None:
                result.append(Finding(
                    severity,
                    self.category,
                    f"{relative(path)}: syntax error: {error}",
                ))
                continue
            wrapper = cst.MetadataWrapper(tree)
            visitor = _BlankLineFinder(relative(path))
            wrapper.visit(visitor)
            result.extend(visitor.findings)
        return result
